File: edx/cs50/C/credit.c
#include <stdio.h>
#include <cs50.h>

//function to sum the digits of given number
long sum_of_digits(long x)
{
    long sum = 0;
    while (x != 0)
    {
        sum += x % 10;
        x = x / 10;
    }
    return (sum);
}

int main(void)
{
    // get number from the user
    long n = get_long("Number: ");
    long sum = 0;
    int product = 0;
    bool odd = true;
    int current_digit = 0;
    int previous_digit = 0;
    int digit_count = 0;

    while (n != 0)
    {
        digit_count++;
        previous_digit = current_digit;
        // mod 10 gets us the rightmost current digit
        current_digit = n % 10;

        // if current digit index is odd then we add it to our sum
        if (odd == true)
        {
            sum += current_digit;
            odd = false;
        }

        // if current digit is even we'll multiply by 2 then add the product of the digits to our sum
        else
        {
            product = current_digit * 2;
            sum += sum_of_digits(product);
            odd = true;
        }
        n = n / 10;
    }

    int index = current_digit * 10 + previous_digit;

    // VISA format
    if (sum % 10 == 0 && current_digit == 4 && (digit_count == 13 || digit_count == 16))
    {
        printf("VISA\n");
    }
    // AMEX format
    else if (sum % 10 == 0 && (index == 34 || index == 37) && digit_count == 15)
    {
        printf("AMEX\n");
    }
    // Mastercard format
    else if (sum % 10 == 0 && (index >= 51 && index <= 55) && digit_count == 16)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: edx/cs50/Python/credit.py
from cs50 import get_int

number = str(get_int("Number: "))


def check(x):
    product_sum = 0
    total_sum = 0
    n = len(x)

    for i in range(n):
        if i % 2 != 0:
            product = str(int(x[n-i-1]) * 2)
            for j in range(len(product)):
                product_sum += int(product[j])
        else:
            total_sum += int(x[n-i-1])

    total_sum += product_sum
    total_sum = str(total_sum)

    if total_sum[len(total_sum)-1] == '0':
        return True
    else:
        return False


if check(number):
    n = len(number)

    if n == 15 and number[0:2] in ('34', '37'):
        print("AMEX\n")
    elif n == 16 and number[0:2] in ('51', '52', '53', '54', '55'):
        print("MASTERCARD\n")
    elif n in (13, 16) and number[0] == '4':
        print("VISA\n")
    else:
        print("INVALID\n")

else:
    print("INVALID\n")

